sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -135,8 +138,15 @@ case class BroadcastNestedLoopJoinExec(
    *
    *   LeftOuter with BuildRight
    *   RightOuter with BuildLeft
+   *   LeftSingle with BuildRight
+   *
+   * For the (LeftSingle, BuildRight) case we pass 'checkMatches' function that
+   * makes sure there is at most 1 matching build row per every probe tuple.
+   * For all other cases, 'checkMatches' is a no-op.
    */
-  private def outerJoin(relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+  private def outerJoin(
+      relation: Broadcast[Array[InternalRow]],
+      checkMatches: Int => Int): RDD[InternalRow] = {
```

Comment:
shouldn't it be `Int => Unit`?

Reply:
it also increments the number of matches and returns the updated count.
Otherwise we would need to increment it in the body of the join and do it for LeftSingle AND LeftOuter joins. This way we avoid incrementing the counter for LeftOuter join where it is not needed.

Reply:
removed the checkMatches in favor of your next suggestion.